### ISSUE TITLE
Calling into C/C++ Code

### ISSUE THREAD

**Issue by athei** [koute/stdweb#242]
Lets assume I have a C library compiled to wasm with emscripten. With javascript I can call into C ABI using `Module.ccall('function_name')`.  I did not test wrapping hat into a `js!` macro but It I guess that not incur some really unnecessary call overhead. 

1. Is there a better way to make a wasm to wasm call, which is cross source language? More specific: How do I call into a C wasm module using stdweb?

2. Can I still use the wasm32-unknown-unknown target for Rust when the C code is compiled with emscripten? Compiling C with the native target is kind of hard because emscripten has browser implementations of common libs (libc++, pthreads).

**Comment by koute** [comment]
> Is there a better way to make a wasm to wasm call, which is cross source language? More specific: How do I call into a C wasm module using stdweb?

Currently you *probably* have to go through JS to do that. In theory I think it should be possible to link multiple `.wasm` files together, but I'm not sure if any such tooling exists, and even if it did then emscripten-style `.wasm` modules expect a different environment than `cargo-web`-style `.wasm` modules and use their own allocators so it would most likely not work. ([Dynamic linking support](https://github.com/WebAssembly/tool-conventions/blob/master/DynamicLinking.md) is supposedly a planned feature.)

> Can I still use the wasm32-unknown-unknown target for Rust when the C code is compiled with emscripten?

Yes. The generated `.js` module is completely encapsulated.

**Comment by athei** [comment]
I agree. emscripten has a huge *.js glue file that does a lot of environment setup. I guess go trough javascript for now. Thanks for your answers.

**Comment by koute** [comment]
You might also want to take a look at [wasm-to-rust](https://github.com/CryZe/wasm-to-rust), which can supposedly (I haven't tried it myself) compile down WASM bytecode to Rust. See [here](https://www.reddit.com/r/programming/comments/8tglyb/c2rust_translate_c_into_rust_code/e17so0h/) for more details.

**Comment by athei** [comment]
Very interesting project. Could be used to import any library into rust if it works.

**Comment by giraffekey** [comment]
Perhaps you can use wasmtime: https://github.com/bytecodealliance/wasmtime

**Comment by giraffekey** [comment]
They use `.wat` for simplicity but I'm sure you can use `.wasm` too

https://docs.wasmtime.dev/examples-rust-gcd.html
